File: 1408-find-the-smallest-divisor-given-a-threshold/solution.py
class Solution:
    def func(self,nums,d):
        summ=0
        for i in nums:
            summ=summ+math.ceil(i/d)
        return summ
    def smallestDivisor(self, nums: List[int], threshold: int) -> int:
        start=1
        end=max(nums)
        while(start<=end):
            mid=start+(end-start)//2
            res=self.func(nums,mid)
            ans=0
            if(res<=threshold):
                ans=mid
                end=mid-1
            else:
                start=mid+1
        return start

File: 1408-find-the-smallest-divisor-given-a-threshold/solution.java
class Solution {
    public int sum(int[] nums,int k){
        int s=0;
        for(int num:nums){
            s=s+(int)(Math.ceil((double)num/k));
        }
        return s;
    }
    public int smallestDivisor(int[] nums, int threshold) {
        int low=1;
        int high=0;
        for(int num:nums){
            high=Math.max(high,num);
        }
        int ans=0;
        while(low<=high){
            int mid=low+(high-low)/2;
            int su=sum(nums,mid);
            if(su<=threshold){
                ans=mid;
                high=mid-1;
            }
            else{
                low=mid+1;
            }
        }
        return ans;
    }
}
